### PR TITLE
feat(logging): add tensorboard support

### PR DESCRIPTION
**Pull request by zzzzwj**
Now support visualization of training loss & validation metrics.
Here is an example:
image

### CODE REVIEW

**Comment by xuyxu** [comment]
Hi @zzzzwj, thanks for the hard work! The visualization looks nice. Could you solve the conflicts in `snapshot_ensemble.py`, so that we can see the changes on the coverage rate of unit tests.

**Comment by xuyxu** [comment]
I have looked at the implementation, the biggest concern is that is there a better way than adding another parameter (i.e.,`tb_logger`) for the `fit` method. For example, can we add another parameter `set_tb_logger` for `__init__`. If it's `True`, we then initialize a tensorboard logger during the instantiation of the ensemble, similar to this [line](https://github.com/xuyxu/Ensemble-Pytorch/blob/master/torchensemble/_base.py#L79).

**Comment by zzzzwj** [comment]
Sure. But there is still a problem that where I can place the parameter `use_tb_logger`. Or shall I discard the option of using tb_logger ?

**Comment by xuyxu** [comment]
> Sure. But there is still a problem that where I can place the parameter `use_tb_logger`. Or shall I discard the option of using tb_logger ?

It could be placed right below `use_cuda` in `__init__`

**Comment by zzzzwj** [comment]
> > Sure. But there is still a problem that where I can place the parameter `use_tb_logger`. Or shall I discard the option of using tb_logger ?
> 
> It could be placed right below `use_cuda` in `__init__`

Okay, that looks great to me.

**Comment by xuyxu** [comment]
In addition, it would be much better to implement a logger class that wraps both the tensorboard logger and the python logger, which will make codes related to logging in each ensemble more succinct. ;-)

**Comment by zzzzwj** [comment]
> In addition, it would be much better to implement a logger class that wraps both the tensorboard logger and the python logger, which will make codes related to logging in each ensemble more succinct. ;-)

I've tried to merge tensorboard & python logging into single file. Now there is no option of `use_tb_logger` in `__init__`. All configs of logging module can be set by function `set_logger`. A logger class is also a good solution. I'll consider on it.

File: requirements.txt [hunk 1]
```diff
@@ -1,3 +1,4 @@
 torch>=1.4.0
 torchvision>=0.2.2
-scikit-learn>=0.23.0
\ No newline at end of file
+scikit-learn>=0.23.0
+tensorboard
```
Comment (xuyxu) [inline requirements.txt:4]: We need to find the minimal version for tensorboard that is compatible with other dependencies.

**Review by xuyxu — COMMENTED** [review]
Please refer to the comment, thanks!

**Comment by xuyxu** [comment]
In addition, here are two problems that may need extra improvements from my point view:

* Add the tensorboard logger in unit tests. Currently, codes on tensorboard logging are not hit by our unit tests.
* Find a better way of logging the training loss for voting/bagging/adversarial_training, currently it's missing, right?

**Comment by xuyxu** [comment]
Nice work @zzzzwj, I will merge the PR once the documentation is updated.

File: requirements.txt (hunk 1 above)
Comment (zzzzwj) [reply requirements.txt:4]: Sorry, I can't tell the minimal version of tensorboard. I've found nothing of version conflict in pytorch's docs and I've tried several version of tensorboard among v1.13 to v2.4, everything works well.

Comment (zzzzwj) [reply requirements.txt:4]: Sorry, I can't tell the minimal version of tensorboard. I've found nothing of version conflict in pytorch's docs and I've tried several version of tensorboard among v1.13 to v2.4, everything works well.

Comment (xuyxu) [reply requirements.txt:4]: How about pinning the version of tensorboard to `2.*`? I have checked the PyPI, the version `2.0.0` was released in Sep, 2019, this should be appropriate for most users.

**Review by xuyxu — COMMENTED** [review]
LGTM.
